### PR TITLE
fix: negation rules not respected in Snyk Code [HEAD-236]

### PR DESCRIPTION
**Pull request by asaf92**
### Description

Fixed a problem where negation rules were not respected in the file filtering of Snyk Code scans.

This PR also contains optimization of the file filtering in Snyk Code scans:
- Files are now streamed via a channel rather than waiting for all the filtered files to be collected before calling `createBundle`
- Ignore rules are being checked in parallel

### Checklist

- [x] Tests added and all succeed

### CODE REVIEW

File: infrastructure/code/code.go [hunk 1]
```diff
@@ -163,17 +157,9 @@ func (sc *Scanner) Scan(ctx context.Context, path string, folderPath string) (is
...
-	folderFiles, err := sc.files(folderPath)
-	if err != nil {
-		log.Warn().
-			Err(err).
-			Str("method", "domain.ide.workspace.Folder.ScanFolder").
-			Str("workspaceFolderPath", folderPath).
-			Msg("error getting workspace files")
-	}
-
-	metrics := sc.newMetrics(len(folderFiles), startTime)
```
Comment (asaf92) [inline infrastructure/code/code.go:175]: Still need to add the file count

File: infrastructure/code/code.go [hunk 2]
```diff
@@ -339,10 +265,18 @@ func (sc *Scanner) handleCreationAndUploadError(path string, err error, msg stri
...
+func isNoFilesError(err error) bool {
+	_, ok := err.(noFilesError)
+	return ok
+}
+
 func (sc *Scanner) createBundle(ctx context.Context,
 	requestId string,
 	rootPath string,
-	filePaths []string,
+	filePaths <-chan string,
```
Comment (bastiandoetsch) [inline infrastructure/code/code.go:279]: Why a channel instead of a slice?

File: infrastructure/code/code.go [hunk 3]
```diff
@@ -163,17 +156,9 @@ func (sc *Scanner) Scan(ctx context.Context, path string, folderPath string) (is
...
-		log.Warn().
-			Err(err).
-			Str("method", "domain.ide.workspace.Folder.ScanFolder").
-			Str("workspaceFolderPath", folderPath).
-			Msg("error getting workspace files")
-	}
-
-	metrics := sc.newMetrics(len(folderFiles), startTime)
-	results, err := sc.UploadAndAnalyze(span.Context(), folderFiles, folderPath, metrics, changedFiles)
+	filesCh := filefilter.FindNonIgnoredFiles(folderPath)
```
Comment (bastiandoetsch) [inline infrastructure/code/code.go:159]: Let's name it `filesChannel`. No need to shorten it here :)

File: infrastructure/filefilter/filter.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	var globs []string
+	if path != f.repoRoot {
+		globs = append(globs, f.globsPerFolder[filepath.Dir(path)]...)
+	} else {
+		defaultGlobs := []string{"**/.git/**", "**/.svn/**", "**/.hg/**", "**/.bzr/**", "**/.DS_Store/**"}
+		globs = append(globs, defaultGlobs...)
+	}
+
+	for _, ignoreFile := range f.ignoreFiles {
+		ignoreFilePath := filepath.Join(path, ignoreFile)
```
Comment (bastiandoetsch) [inline infrastructure/filefilter/filter.go:104]: where do we report the progress of the filewalking?

File: infrastructure/code/code.go [hunk 5]
```diff
@@ -207,77 +192,18 @@ func (sc *Scanner) waitForScanToFinish(scanStatus *ScanStatus, folderPath string
...
-				log.Debug().
-					Str("method", "domain.ide.workspace.Folder.Files").
-					Str("path", path).
-					Err(err).
-					Msg("error traversing files")
-				return nil
-			}
-			filesWalked++
-			percentage := math.Round(float64(filesWalked) / float64(fileCount) * 100)
-			t.ReportWithMessage(
```
Comment (bastiandoetsch) [inline infrastructure/code/code.go:242]: this seems to be missing in the new implementation (but is important for user progress reporting on big repos).

File: infrastructure/code/code.go [hunk 6]
```diff
@@ -207,77 +192,18 @@ func (sc *Scanner) waitForScanToFinish(scanStatus *ScanStatus, folderPath string
...
-
-func (sc *Scanner) newMetrics(fileCount int, scanStartTime time.Time) *ScanMetrics {
+func (sc *Scanner) newMetrics(scanStartTime time.Time) *ScanMetrics {
 	if scanStartTime.IsZero() {
 		scanStartTime = time.Now()
 	}
 
 	return &ScanMetrics{
 		lastScanStartTime: scanStartTime,
-		lastScanFileCount: fileCount,
```
[the remark this answers is not in the document]
Comment (asaf92) [reply infrastructure/code/code.go:275]: I added the file count after the `createBundle` call

File: infrastructure/filefilter/filter.go (hunk 4 above)
Comment (asaf92) [reply infrastructure/filefilter/filter.go:104]: We can't report the progress percentage now. Previously we had one walk where we collect all the ignore rules and count the files, and then a 2nd walk where we filter the files. The progress of the 2nd walk was reported and the progress percentage was the count of the walked files divided by the total file count. 

Now we only have one walk, so we can't really know how many files are left.

We still have other parts where progress is reported, for example when uploading/analyzing/etc

File: infrastructure/code/code_test.go [hunk 7]
```diff
@@ -334,33 +346,6 @@ func TestUploadAndAnalyze(t *testing.T) {
 	)
 }
 
-func Test_GetWorkspaceFolderFiles(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (asaf92) [reply infrastructure/code/code_test.go:337]: Yes, I added tests for this and other ignore cases and made the tests parameterized

File: infrastructure/filefilter/filter.go (hunk 4 above)
Comment (bastiandoetsch) [reply infrastructure/filefilter/filter.go:104]: Then we need at least a progress message that is unquantifiable that says "Collecting files"

File: infrastructure/filefilter/filter.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+	repoRoot               string
+	ignoreFiles            []string
+	globsPerFolder         map[string][]string
+	ignoreCheckerPerFolder map[string]ignore.IgnoreParser
+}
+
+func newRepoIgnoresFilter(rootFolder string) *repoIgnoresFilter {
+	return &repoIgnoresFilter{
+		repoRoot:               rootFolder,
+		ignoreFiles:            []string{".gitignore", ".dcignore", ".snyk"},
```
Comment (michelkaporin) [inline infrastructure/filefilter/filter.go:30]: This set should be dictated by `/filters` endpoint.

**Comment by michelkaporin** [comment]
@asaf92 I love this PR but let's not deploy this until we stay at 100% rollout for Snyk Code for some time to derisk any complications.

File: infrastructure/filefilter/filter.go (hunk 8 above)
Comment (bastiandoetsch) [reply infrastructure/filefilter/filter.go:30]: Not really - this is not Snyk Code specific, even though it's used by Snyk Code.

File: infrastructure/filefilter/filter.go [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+func (f *repoIgnoresFilter) findNonIgnoredFiles() <-chan string {
+	resultsCh := make(chan string)
+	var wg sync.WaitGroup
+	go func() {
+		defer close(resultsCh)
+		// When walking a directory, the iteration is the directory itself (`path` parameter is the directory path).
+		// The ignore files are immediately parsed, and the iterations of the files come after the ignore rules have been loaded
+		err := filepath.WalkDir(f.repoRoot, func(path string, dirEntry os.DirEntry, err error) error {
+			if err != nil {
+				return filepath.SkipAll
```
[the remark this answers is not in the document]
Comment (asaf92) [reply infrastructure/filefilter/filter.go:45]: Changed to skip dir + log error

File: infrastructure/filefilter/filter.go (hunk 4 above)
Comment (asaf92) [reply infrastructure/filefilter/filter.go:104]: Done

File: infrastructure/filefilter/filter.go [hunk 10]
```diff
@@ -29,6 +31,7 @@ func newFileFilter(rootFolder string) *fileFilter {
 		ignoreFiles:            []string{".gitignore", ".dcignore", ".snyk"},
 		globsPerFolder:         make(map[string][]string),
 		ignoreCheckerPerFolder: make(map[string]ignore.IgnoreParser),
+		logger:                 log.With().Str("component", "fileFilter").Str("repoRoot", rootFolder).Logger(),
```
Comment (michelkaporin) [inline infrastructure/filefilter/filter.go:34]: That is very neat 👍 Goes into my Golang arsenal

File: licenses/golang.org/x/exp/LICENSE
Comment (michelkaporin) [inline licenses/golang.org/x/exp/LICENSE:1]: Why was it removed, when it's still a dependency?

**Comment by michelkaporin** [comment]
@asaf92 Didn't see the auto-merge enabled, please don't forget to address the comments.

File: licenses/golang.org/x/exp/LICENSE
Comment (asaf92) [reply licenses/golang.org/x/exp/LICENSE:1]: It was used in 1 line in a file I removed (`slices.Contains(...)`)

The license files are automatically added/removed

**Pull request merged**